ts-sdk/src/cosmos/extended-cw-client.ts

```diff
@@ -0,0 +1,61 @@
+import { CosmWasmClient, type HttpEndpoint } from "@cosmjs/cosmwasm-stargate"
+import { Tendermint34Client } from "@cosmjs/tendermint-rpc"
+import axios from "axios"
+
+export class ExtendedCosmWasmClient extends CosmWasmClient {
+  private restUrl!: string
+
+  static async connectWithHeightSupport(endpoint: HttpEndpoint | string, restEndpoint: string) {
+    const tmClient = await Tendermint34Client.connect(endpoint)
+    const client = new ExtendedCosmWasmClient(tmClient)
+
+    client.restUrl = restEndpoint.replace(/\/+$/, "")
+    return client
+  }
+
+  async queryContractSmartAtHeight(
+    contract: string,
+    queryMsg: Record<string, unknown>,
+    height: number,
+  ) {
+    const base = this.restUrl
+    const encoded = btoa(JSON.stringify(queryMsg))
+    const url = `${base}/cosmwasm/wasm/v1/contract/${contract}/smart/${encoded}`
+
+    const resp = await axios.get(url, {
+      headers: {
+        "Content-Type": "application/json",
+        "x-cosmos-block-height": height.toString(),
+      },
+    })
+    if (resp.status < 200 || resp.status >= 300) {
+      throw new Error(`HTTP ${resp.status}: ${JSON.stringify(resp.data)}`)
+    }
+
+    return resp.data
+  }
+
+  async getBalanceAtHeight(address: string, denom: string, height: number) {
+    const base = this.restUrl
+    const url = `${base}/cosmos/bank/v1beta1/balances/${address}`
+    const resp = await fetch(url, {
```

Comment:
i recommend not going for this approach of extending `CosmWasmClient`, at all and to instead just write simple functions that make the Http call and return an Effect with proper error types rather than throwing errors. Using the client here is basically  useless because all that we're doing is just extracting the restURL from the client and then making a simple http request. Because we're doing it this way we lose valuable error information as we're not using our Effectful methods of making http calls that we have elsewhere in the sdk